### PR TITLE
Issue 19220 function parameter arity

### LINKED ISSUE #19220: Expected 1-3 arguments, but got 2
**TypeScript Version:**  2.6.0-dev.20171015

**Code**

```ts
declare function f(x: number): number;
declare function f(x: number, y: number, z: number): number;

f(1, 2);
```

**Expected behavior:**

`Expected 1 or 3 arguments, but got 2.`

**Actual behavior:**

`src/a.ts(4,1): error TS2554: Expected 1-3 arguments, but got 2.`

### PR DESCRIPTION
**Pull request by rflorian**
Fixes #19220 

Further changes could be made to rigorously check all available overload signatures in order to identify cases where emitting possible overloads as a range is applicable and thus respond with more accurate diagnostics.

### CODE REVIEW

**Comment by rflorian** [comment]
I am currently working on further analyzing available signatures to convey for more useful error information in complex cases.

**Comment by rflorian** [comment]
After further investigating all possible cases of function parameter arity mismatches I will most likely include the following new error messages in this PR:

```
TS2571: "Expected no more than {0} arguments, but got {1}."
For cases such as:
declare function f1();
declare function f1(a: number, b: number);
f1(1, 2, 3); // -> TS2571 with "2", "3"

TS2572: "No overload expects {0} arguments. The most likely overloads that match expect either {1} arguments or at least {2} arguments."
For cases such as:
declare function f2();
declare function f2(a: number);
// and either non-contiguous declarations requiring more parameters
declare function f2(a: number, b: number, c: number);
declare function f2(a: number, b: number, c: number, d: number, e: number);
// or at least one declaration which has a rest parameter
declare function f2(a: number, b: number, c: number, ...d: number[]);
f2(1, 2); // -> TS2572 with "2", "0-1", "3"

TS2573: "No overload expects {0} arguments. The most likely overloads that match expect either up to {1} arguments or {2} arguments."
For cases such as:
declare function f3();
declare function f3(a: number, b: number);
declare function f3(a: number, b: number, c: number, d: number);
declare function f3(a: number, b: number, c: number, d: number, e: number);
f3(1, 2, 3); // -> TS2573 with "3", "2", "4-5"

TS2574: "No overload expects {0} arguments. The most likely overloads that match expect either up to {1} arguments or at least {2} arguments."
For cases such as:
declare function f4();
declare function f4(a: number, b: number);
// and either non-contiguous declarations requiring more parameters
declare function f4(a: number, b: number, c: number, d: number);
declare function f4(a: number, b: number, c: number, d: number, e: number, f: number);
// or at least one declaration which has a rest parameter
declare function f4(a: number, b: number, c: number, d: number, ...e: number[]);
f4(1, 2, 3); // -> TS2574 with "3", "2", "4"
```

Feedback on these proposed error conditions and messages would be appreciated. 

Cases where the expected number of arguments of both signatures accepting less and more arguments than which are provided to the faulty function call will be handled through the existing error TS2554 by passing in 2 range parameters.

File: tests/cases/fourslash/signatureHelpOnOverloadsDifferentArity3.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ verify.currentSignatureParameterCountIs(2);
 verify.currentParameterHelpArgumentNameIs("b");
 verify.currentParameterSpanIs("b: boolean");
 
-edit.insert("x, ");
+// edit.insert("x, "); // TODO: This block looks like a duplicate of the above code
```
Comment (rflorian) [inline tests/cases/fourslash/signatureHelpOnOverloadsDifferentArity3.ts]: Is this test case necessary?

**Comment by mhegazy** [comment]
@sandersn can you please review this change.

**Comment by rflorian** [comment]
@sandersn This PR is now up to date with master again.

**Comment by sandersn** [comment]
@rflorian Just a brief note. I started looking at the PR.

However, I don't believe this is the right solution. I believe that it catches a case (discontinuous overload sets) where the target function is wrong, not the calling code, and so the error message should (1) point the user to the target function (2) be much easier to implement, since it's a rare case.

I don't have data to back me up yet, but I'm in the middle of making a version of the compiler that fails whenever it sees a discontinuous overload set, after which I'll run it on Definitely Typed to see whether real authors ever intentionally write them. I'll report back what I find.

**Comment by sandersn** [comment]
OK, I've got results from the experiment. Turns out I was wrong. The DOM, jquery and node all have discontinuous overload arities, as do many other projects, including Typescript itself. A few of these I looked at are just fussy, but do aid correctness. The others all follow a similar pattern: Sets of parameters get added in logical groups, such as x/y pairs. An example is `(sourceX, sourceY); (sourceX, sourceY, targetX, targetY)`, where the second overload lets you specify both source and target locations.

So I do think this change is necessary. There is still a trade off between smart messages backed by very complex code versus a simple message backed by simple code. I'm not sure where we should end up on that; I'll take a closer look at the code to help me decide.

In case anybody wants to replicate the experiment, here is the code:

```ts
        function kaboom(signatures: ReadonlyArray<Signature>) {
            if (length(signatures) === 0) return;
            let spans = [];
            let min = Number.POSITIVE_INFINITY;
            let max = Number.NEGATIVE_INFINITY;
            for (const sig of signatures) {
                let x = { min: getMinArgumentCount(sig), max: getParameterCount(sig) }
                spans.push(x);
                min = Math.min(min, x.min);
                max = Math.max(max, x.max);
            }
            for (var i = min; i <= max; i++) {
                if (!find(spans, ({min,max}) => min <= i && i <= max)) {
                    console.log("Couldn't find " + i + " at " + getTextOfNode(signatures[0].declaration!));
                }
            }
        }

// ... early in checkFunctionOrMethodDeclaration ...
            const t = getTypeOfSymbol(getSymbolOfNode(node))
            const signatures = getSignaturesOfType(t, SignatureKind.Call);
            kaboom(signatures);
```

It's slow and doesn't handle rest parameters quite right but is good enough to find lots of examples.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -19043,26 +19043,48 @@ namespace ts {
                 diagnostics.add(getTypeArgumentArityError(node, signatures, typeArguments));
             }
             else if (args) {
-                let min = Number.POSITIVE_INFINITY;
-                let max = Number.NEGATIVE_INFINITY;
```
Comment (sandersn) [inline src/compiler/checker.ts]: I think you should revert all the first section of this change and add two more variables. This worked for me in simple tests:

```ts
let belowArgCount = Number.NEGATIVE_INFINITY
let aboveArgCount = Number.POSITIVE_INFINITY
for (const sig of signatures) {
    minCount = getMinArgumentCount(sig)
    maxCount = getParameterCount(sig)
    if (minCount < argCount && argCount - minCount < argCount - belowArgCount) belowArgCount = minCount
    if (argCount < maxCount && maxCount - argCount < aboveArgCount - argCount) aboveArgCount = maxCount
    min = Math.min(min, minCount)
    max = Math.max(max, maxCount)
}
```

(Note that you'll have to calculate argCount earlier.) Then you can use belowArgCount/aboveArgCount in the new error reporting branch you added in the second half of this change.

File: tests/baselines/reference/genericRestParameters1.errors.txt [hunk 3]
```diff
@@ -26,7 +26,7 @@ tests/cases/conformance/types/rest/genericRestParameters1.ts(31,1): error TS2556
     f1(ns[0], ns[1], true);
     f1(...ns, true);  // Error, tuple spread only expanded when last
     ~~~~~~~~~~~~~~~
-!!! error TS2556: Expected 3 arguments, but got 1 or more.
+!!! error TS2557: Expected at least 3 arguments, but got 1 or more.
```
Comment (rflorian) [inline tests/baselines/reference/genericRestParameters1.errors.txt]: These still emit a slightly changed error message? Is the changed behavior in this case wanted and/or acceptable?

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -2056,6 +2056,10 @@
         "category": "Error",
         "code": 2574
     },
+    "No overload expects {0} arguments, but overloads do exist that expect either {1} or at least {2} arguments.": {
```
Comment (rflorian) [inline src/compiler/diagnosticMessages.json]: I added your simplified error message, however I feel like it might hide information from the user in cases where discontinuous overload sets with argument counts less than the current argument count exist.

**Comment by rflorian** [comment]
@sandersn Thank you for the detailed analysis and hints on the PR. I incorporated your suggestions into the PR with slight modifications. I do not fully understand the requirements for diagnostics generation with regards to the new new tuple-spreading code so those might not work fully as intended.

**Comment by rflorian** [comment]
When checking for rest parameters, hasEffectiveRestParameter is now properly utilized again. This fixed both cases where tuple unpacking parameters were erroneously assessed as rest parameters.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -19048,24 +19048,37 @@ namespace ts {
...
-                if (argCount <= max && hasSpreadArgument) {
-                    argCount--;
-                }
-                const error = hasRestParameter && hasSpreadArgument ? Diagnostics.Expected_at_least_0_arguments_but_got_1_or_more :
-                    hasRestParameter ? Diagnostics.Expected_at_least_0_arguments_but_got_1 :
-                    hasSpreadArgument ? Diagnostics.Expected_0_arguments_but_got_1_or_more :
-                    Diagnostics.Expected_0_arguments_but_got_1;
-                diagnostics.add(createDiagnosticForNode(node, error, paramCount, argCount));
+                const hasSpreadArgument = getSpreadArgumentIndex(args) > -1;
+                if (argCount <= max && hasSpreadArgument) argCount--;
```
Comment (sandersn) [inline src/compiler/checker.ts]: Please revert this change too. (I'm not sure whether the linter will let it through anyway.)

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -19048,24 +19048,37 @@ namespace ts {
...
+
+                if (hasRestParameter || hasSpreadArgument) {
+                    const error = hasRestParameter && hasSpreadArgument ? Diagnostics.Expected_at_least_0_arguments_but_got_1_or_more :
+                        hasRestParameter ? Diagnostics.Expected_at_least_0_arguments_but_got_1 : Diagnostics.Expected_0_arguments_but_got_1_or_more;
```
Comment (sandersn) [inline src/compiler/checker.ts]: Please revert this line-join and put `Diagnostics.Expected 0 arguments but got 1 or more` back on its own line.

**Comment by rflorian** [comment]
Sounds good, I believe everything is now formatted as expected.

**Comment by sandersn** [comment]
Thanks for all your work on this one @rflorian !

**Comment by rflorian** [comment]
Thanks for helping me with the PR @sandersn !
